### DISCUSSION TITLE
Can't get mesh picking to work with modified transform on model

### DISCUSSION THREAD

**Discussion by Pirolye** [raysan5/raylib#2675] [Q&A]
Hi,

I have a struct which contains a model. The struct has a function to update the transform of the model so I don't have to do it manually all the time.

I also have a mesh picking system wherein the world struct sends out a request to all entity structs to check against their own models or anything that I want to include in the selection process. 

This works perfectly for models that have { 0 } transforms. Anything else doesn't work.

The collision check is failing at line 3683 of rmodels.c ( `if ((u < 0.0f) || (u > 1.0f)) return collision;` ) 

Transform change algorithm: (enttTransform is a struct with 2 Vector3s and a float, not connected to anything)
`

	cubeModel.transform = MatrixMultiply(cubeModel.transform, MatrixScale(enttTransform.scale, enttTransform.scale, enttTransform.scale));
	cubeModel.transform = MatrixMultiply(cubeModel.transform, MatrixRotateXYZ(Vector3{ DEG2RAD * enttTransform.rot.x, DEG2RAD * enttTransform.rot.y, DEG2RAD * enttTransform.rot.z }));
	cubeModel.transform = MatrixMultiply(cubeModel.transform, MatrixTranslate(enttTransform.pos.x, enttTransform.pos.y, enttTransform.pos.z));

`

Here's how the entity in question deals with the pick request: (This works fine)

`

	RayCollision meshHitInfo = { 0 };
	for (int m = 0; m < cubeModel.meshCount; m++)
	{
		// NOTE: We consider the model.transform for the collision check but 
		// it can be checked against any transform Matrix, used when checking against same
		// model drawn multiple times with multiple transforms
		meshHitInfo = GetRayCollisionMesh(inRay, cubeModel.meshes[m], cubeModel.transform); // FAILING ON NON-ZERO TRANSFORMS
		if (meshHitInfo.hit)
		{
			
			return this;

			break;  // Stop once one mesh collision is detected, the colliding mesh is m
		}
		else
		{
			return nullptr;
		}
	}

`

I have gone through step-by-step all 3 algorithms (transform change -> collision check request -> rmodels.c collision check, although I don't understand that one)

Here's how the model's transform looks after the change algortihm: (Scaled down and moved a bit)
![image](https://user-images.githubusercontent.com/22526998/187444763-be1b3869-19f4-4c1e-b728-1ae5a0f66c98.png)

It would be greatly appreciated if someone could nudge me in the correct direction,
Pirolye

**Comment by chriscamacho** [comment]
just a thought, are you setting the transform to identity?  as

`cubeModel.transform = MatrixMultiply(cubeModel.transform, MatrixScale(enttTransform.scale, enttTransform.scale, enttTransform.scale));`

will be picking up the previous state of the transform...

**Reply by Pirolye** [reply]
I'm afraid I don't understand how to do that. The docs don't have anything and I can't decode it from the headers...

**Reply by Pirolye** [reply]
Update 1: I now understand the suggestion. I won't worry about that for now.

Update 2: It looks like ray's algorithm is wonky as from certain positions I am able to select the cube with the modified transform. Checked the code again, the transform is applied and passed down correctly everywhere.

**Reply by Pirolye** [reply]
![image](https://user-images.githubusercontent.com/22526998/187472510-e84d74fa-441d-491f-83c8-268ff2d43f20.png)
This is the current standing. Please ignore the red lines, they are not used anymore.

**Reply by Pirolye** [reply]
Update 3: I tried to manually update bounding box positions since mesh collisions are broken. I tried (+variations) these 2 algorithms:

`


	BoundingBox entityBox = BoundingBox
	{ Vector3 
		{
			enttTransform.pos.x - enttTransform.scale / 2,
			enttTransform.pos.y - enttTransform.scale / 2,
			enttTransform.pos.z - enttTransform.scale / 2
        },
		Vector3 
		{
			enttTransform.pos.x + enttTransform.scale / 2,
			enttTransform.pos.y + enttTransform.scale / 2,
			enttTransform.pos.z + enttTransform.scale / 2
		} 
	};
	
	selectionBoundingBox = entityBox;
	
	/*
	
	selectionBoundingBox.min = Vector3Scale(selectionBoundingBox.min, enttTransform.scale * enttTransform.scale);
	selectionBoundingBox.max = Vector3Scale(selectionBoundingBox.max, enttTransform.scale * enttTransform.scale);



	//Vector3 bruh = { enttTransform.scale / 2.0f, enttTransform.scale / 2.0f, enttTransform.scale / 2.0f };
	Vector3 bruh = { enttTransform.scale, enttTransform.scale, enttTransform.scale };
	
	Vector3 posScaled = Vector3Scale(enttTransform.pos, (enttTransform.scale * enttTransform.scale));

	Vector3 aaa = Vector3{ posScaled.x + posScaled.x + posScaled.x, posScaled.y + posScaled.y + posScaled.y, posScaled.z + posScaled.z + posScaled.z };
	selectionBoundingBox.min = Vector3Add(selectionBoundingBox.min, aaa);
	selectionBoundingBox.max = Vector3Add(selectionBoundingBox.max, aaa);
	//Vector3 bbb = Vector3Subtract(aaa, bruh);

	*/



`
